tsconfig.base.json

```diff
@@ -0,0 +1,45 @@
+{
+	// This tsconfig file contains the shared config for the build (tsconfig.build.json) and type checking (tsconfig.json) config.
+	"include": [],
+	"compilerOptions": {
+		// Incremental builds
+		// NOTE: Enabling incremental builds speeds up `tsc`. Keep in mind though that it does not reliably bust the cache when the `tsconfig.json` file changes.
+		"incremental": false,
+
+		// Type checking
+		"strict": true,
+		"useDefineForClassFields": true, // Not enabled by default in `strict` mode unless we bump `target` to ES2022.
+		"noFallthroughCasesInSwitch": true, // Not enabled by default in `strict` mode.
+		"noImplicitReturns": true, // Not enabled by default in `strict` mode.
+		"useUnknownInCatchVariables": true, // TODO: This would normally be enabled in `strict` mode but would require some adjustments to the codebase.
+		"noImplicitOverride": true, // Not enabled by default in `strict` mode.
+		"noUnusedLocals": true, // Not enabled by default in `strict` mode.
+		"noUnusedParameters": true, // Not enabled by default in `strict` mode.
+		"exactOptionalPropertyTypes": true,
+		// TODO: Uncomment and fix types.
+		// "noUncheckedIndexedAccess": true,
+
+		// JavaScript support
+		"allowJs": false,
+		"checkJs": false,
+
+		// Interop constraints
+		"esModuleInterop": false,
+		"allowSyntheticDefaultImports": false,
+		"forceConsistentCasingInFileNames": true,
+		"verbatimModuleSyntax": false,
+		"importHelpers": true, // This is only used for build validation. Since we do not have `tslib` installed, this will fail if we accidentally make use of anything that'd require injection of helpers.
+
+		// Language and environment
+		"moduleResolution": "NodeNext",
```

Comment:
This is copied from the Viem project: https://github.com/wevm/viem

Reply:
o pretty substantial, should we put in their license too?

Reply:
> Configuration files are not typically copyrightable as they are used to configure parameters and settings rather than express creative content. They are usually considered functional and utilitarian in nature rather than original works of authorship.

Reply:
whose quote

Reply:
https://github.com/babel/babel/blob/main/packages/babel-preset-typescript/package.json#L10

Reply:
The quote is a summary from GPT which lines up with my experience, a `tsconfig.json` file doesn't really meet the threshold of originality required for copyright protection, in the same way that a shopping list or recipe doesn't.

Reply:
oh that's referring to copyright not protecting that abstract configuration. they don't own the combination of enabling `strict` and setting `moduleResolution` to nodenext. this is file copied, and copyright is meant to protect fixed representations like that

Reply:
I'd like to leave my side of this copyright discussion there, and I'm glad we shared our perspectives on this. I'd like to add some other opinions not related to copyright:

1. I want readers to know that Viem is to credit for this configuration regardless of licensing requires it or not
2. I want our developers to know what this configuration is based on, in case we want to watch for upstream changes that we like too

Reply:
Yup lets add a permalink to the original viem `tsconfig.json` and call it a day.

Reply:
Re: "NodeNext" vs "nodenext", the JSONSchema referenced in the documentation at https://www.typescriptlang.org/docs/handbook/tsconfig-json.html uses `NodeNext`:
 * https://github.com/SchemaStore/schemastore/blob/6dd10d986ab375d7df73dce10b1cc3a7c285a63c/src/schemas/json/tsconfig.json#L421

Reply:
:eyes: woa thanks microsoft